src/agent/testresultreader.ts

```diff
@@ -0,0 +1,404 @@
+var Q = require('q');
+
+export class JUnitResultReader implements trp.IResultReader {
+
+    public readResults(file: string, runContext: trp.TestRunContext) : Q.Promise<ifm.TestRun2> {
+        var defer = Q.defer(); 
+        var _this = this;
+
+        utilities.readFileContents(file, "utf-8").then(function (contents) {
+            var xmlContents = contents.replace("\ufeff", ""); //replace BOM if exits to avoid xml read error
+            return _this.readJUnitTestRunData(xmlContents, runContext);
+        }).then(function (testRun) {
+            defer.resolve(testRun);
+        }).fail(function (err) {
+            defer.reject(err);
+        });        
+
+        return defer.promise;
+    }
+
+    private readJUnitTestRunData(contents: string, runContext: trp.TestRunContext) : Q.Promise<ifm.TestRun2> {
+        var defer = Q.defer(); 
+      
+        var testRun2 : ifm.TestRun2;
+        var _this = this;
+
+        xmlreader.read(contents, function (err, res) {
+            if(err) {
+                defer.resolve(err);
+            } 
+            else {
+                try {
+                    testRun2 = _this.parseJUnitXml(res, runContext);
+                    defer.resolve(testRun2);
+                }
+                catch(ex) {
+                    defer.reject(ex);
+                }
+            }
+        });
+
+        return defer.promise;
+    }
+
+    private parseJUnitXml(res, runContext) {
+        var testRun2 : ifm.TestRun2;
+
+        var buildId = runContext.buildId;
+        var buildRequestedFor = runContext.requestedFor;
+        var platform = "";
+        var config = "";
+
+        //read test run summary - runname, host, start time, run duration
+        var runName = "JUnit";
+        var hostName = "";
+        var timeStamp = new Date(); 
+        var totalRunDuration = 0;
+        var totalTestCaseDuration = 0;
+
+        var rootNode = res.testsuite.at(0);
+
+        if(rootNode) {
+
+            if(rootNode.attributes().name) {
+                runName = rootNode.attributes().name;
+            }
+
+            if(rootNode.attributes().hostname) {
+                hostName = rootNode.attributes().hostname;
+            }
+
+            //assume runtimes from xl are current local time since timezone information is not in the xml. If xml date > current local date, fall back to local
+            if(rootNode.attributes().timestamp) {
+                var timestampFromXml = new Date(rootNode.attributes().timestamp);
+                if(timestampFromXml < new Date()) {
+                    timeStamp = timestampFromXml;
+                }                    
+            }
+
+            if(rootNode.attributes().time) {
+                totalRunDuration = rootNode.attributes().time; //in seconds
+            }
+
+            //find test case nodes in JUnit result xml
+            var testResults = [];
+
+            for(var i = 0; i < rootNode.testcase.count(); i ++) {
```

Comment:
not a big deal, but if there is a foreach variant, may read better compared to index based looping.